src/main/java/io/cryostat/recordings/RecordingArchiveHelper.java

```diff
@@ -39,90 +39,216 @@
 import io.cryostat.core.net.JFRConnection;
 import io.cryostat.core.sys.Clock;
 import io.cryostat.core.sys.FileSystem;
 import io.cryostat.net.ConnectionDescriptor;
 import io.cryostat.net.TargetConnectionManager;
 import io.cryostat.net.reports.ReportService;
+import io.cryostat.net.web.WebServer;
 import io.cryostat.platform.PlatformClient;
+import io.cryostat.rules.ArchivePathException;
+import io.cryostat.rules.ArchivedRecordingInfo;
 import io.cryostat.util.URIUtil;
 
+import org.apache.commons.codec.binary.Base32;
+
 public class RecordingArchiveHelper {
 
     private final TargetConnectionManager targetConnectionManager;
     private final FileSystem fs;
+    private final Provider<WebServer> webServerProvider;
+    private final Logger logger;
     private final Path recordingsPath;
     private final Clock clock;
     private final PlatformClient platformClient;
     private final ReportService reportService;
 
     RecordingArchiveHelper(
             FileSystem fs,
+            Provider<WebServer> webServerProvider,
+            Logger logger,
             @Named(MainModule.RECORDINGS_PATH) Path recordingsPath,
             TargetConnectionManager targetConnectionManager,
             Clock clock,
             PlatformClient platformClient,
             ReportService reportService) {
         this.fs = fs;
+        this.webServerProvider = webServerProvider;
+        this.logger = logger;
         this.recordingsPath = recordingsPath;
         this.targetConnectionManager = targetConnectionManager;
         this.clock = clock;
         this.platformClient = platformClient;
         this.reportService = reportService;
     }
 
-    public String saveRecording(ConnectionDescriptor connectionDescriptor, String recordingName)
-            throws Exception {
+    public Future<String> saveRecording(
+            ConnectionDescriptor connectionDescriptor, String recordingName) throws Exception {
+
+        CompletableFuture<String> future = new CompletableFuture<>();
+
+        try {
+            String saveName =
+                    targetConnectionManager.executeConnectedTask(
+                            connectionDescriptor,
+                            connection -> {
+                                Optional<IRecordingDescriptor> descriptor =
+                                        this.getDescriptorByName(connection, recordingName);
+
+                                if (descriptor.isPresent()) {
+                                    return writeRecordingToDestination(
+                                            connection, descriptor.get());
+                                } else {
+                                    throw new RecordingNotFoundException(recordingName);
+                                }
+                            });
+            future.complete(saveName);
+        } catch (RecordingNotFoundException e) {
+            future.completeExceptionally(e);
+        }
+
+        return future;
```

Comment:
This method could be simplified somewhat since the `executeConnectedTask` actually runs synchronously, but the way you've written it lends itself well to the connected tasks possibly being done asynchronously as well, so I'm okay with leaving it as-is.

However, I think it's a bit awkward for consumers of this internal API to receive a `Future` but also have to catch the generic `Exception` when calling the method. I think it would be less surprising if the `catch` block in this method caught all of those `Exceptions, removing it from the method signature, and pushing them into the `future.completeExceptionally` instead. Does that make sense?